### ISSUE TITLE
[slot_attention] Effective dimensions of LayerNormalization

### ISSUE THREAD

**Issue by vadimkantorov** [google-research/google-research#548]
@tkipf @dosovits, In https://github.com/google-research/google-research/blob/0c1bbe5/slot_attention/model.py#L43-L45:
```python
self.norm_inputs = layers.LayerNormalization()
self.norm_slots = layers.LayerNormalization()
self.norm_mlp = layers.LayerNormalization()
```

With default arguments of [LayerNormalization](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LayerNormalization), this means that implicitly `axis=-1`, `center=True`, `scale=True`

Does it mean that for it's equivalent (in PyTorch lingo) to:
```python
class SlotAttentionLayerNorm(torch.nn.Module):
   def __init__(self, inputs_size, epsilon = 1e-3):
     super().__init__()
     self.gamma = torch.nn.Parameter(torch.ones(inputs_size))
     self.beta = torch.nn.Parameter(torch.zeros(inputs_size))
     self.epsilon = epsilon

  def forward(self, inputs):
     # inputs` has shape [batch_size, num_inputs, inputs_size].
     var, mean = torch.var_mean(inputs, dim = -1, keepdim = True)
     return (inputs - mean) / (var + self.epsilon).sqrt() * self.gamma + self.beta
```

Ironically, this seems to match what PyTorch LayerNorm does, but it doesn't match what LayerNorm is supposed to be doing (or even InstanceNorm for that matter). i.e. computing stats per whole batch instance `batch[b, ...]` :)
![image](https://user-images.githubusercontent.com/1041752/106493465-746d9c80-64b9-11eb-992d-70b0323696cc.png)

If the above code is true, it's computing stats per `batch[b, h, w]` and standardizes the embedding vector themselves (does it even make sense?).

Thanks!

**Comment by vadimkantorov** [comment]
So far my conclusion is that PyTorch's nn.LayerNorm does not perform the LayerNorm (as pictured in GroupNorm paper), and is quite hard to coerce to do so, since it requires static `normalized_shape` and blocks dynamic spatial/temporal batch sizes. TensorFlow is much more flexible and just asks the user to supply the list of dimensions to normalize over. However, by default TF also does not do LayerNorm. In simples case of BTC layout, TF's default LayerNormalization and PyTorch nn.LayerNorm do the same thing and it's not LayerNorm, but something else :(

**Comment by tkipf** [comment]
Your implementation looks correct to me on a quick skim. But couldn't you achieve the same by only passing an integer `slot_size` to `torch.nn.LayerNorm(normalized_shape=slot_size)`?

**Comment by vadimkantorov** [comment]
Yes, and that's what I'm doing that ported code. I was just trying to understand the docs of nn.LayerNorm wrt expected layout (I was worried that TF has default BTC layout, and PyTorch BCT) and the GroupNorm paper visualization confused me (since it also normalizes over the spatial/temporal dimension). The docs phrasing is not very clear wrt about what's averaged "over" versus "per" what the stats are calculated. The TF phrasing is a little better, but also not 100% great.

The original LayerNorm paper is also hard to grasp wrt what exactly is averaged over :(

**Comment by vadimkantorov** [comment]
It seems that in reality, nn.LayerNorm defaults and LayerNormalization defaults to:
![image](https://user-images.githubusercontent.com/1041752/106501534-94a25900-64c3-11eb-865e-8aca5d494162.png)
